pulsar-broker/src/main/java/org/apache/pulsar/broker/delayed/BucketDelayedDeliveryTracker.java

```diff
@@ -0,0 +1,569 @@
+        this.snapshotSegmentLastIndexTable = HashBasedTable.create();
+
+        this.bucketSnapshotStorage = bucketSnapshotStorage;
+
+        this.numberDelayedMessages = 0L;
+
+        this.lastMutableBucketState = new BucketState(-1L, -1L);
+    }
+
+    private void moveScheduledMessageToSharedQueue(long cutoffTime) {
+        TripleLongPriorityQueue priorityQueue = getPriorityQueue();
+        while (!priorityQueue.isEmpty()) {
+            long timestamp = priorityQueue.peekN1();
+            if (timestamp > cutoffTime) {
+                break;
+            }
+
+            long ledgerId = priorityQueue.peekN2();
+            long entryId = priorityQueue.peekN3();
+            sharedBucketPriorityQueue.add(timestamp, ledgerId, entryId);
+
+            priorityQueue.pop();
+        }
+    }
+
+    @Override
+    public void run(Timeout timeout) throws Exception {
+        synchronized (this) {
+            if (timeout == null || timeout.isCancelled()) {
+                return;
+            }
+            moveScheduledMessageToSharedQueue(getCutoffTime());
+        }
+        super.run(timeout);
+    }
+
+    private Optional<BucketState> findBucket(long ledgerId) {
+        if (immutableBuckets.asMapOfRanges().isEmpty()) {
+            return Optional.empty();
+        }
+
+        Range<Long> span = immutableBuckets.span();
+        if (!span.contains(ledgerId)) {
+            return Optional.empty();
+        }
+        return Optional.ofNullable(immutableBuckets.get(ledgerId));
+    }
+
+    private long getBucketId(BucketState bucketState) {
+        Optional<Long> bucketIdOptional = bucketState.getBucketId();
+        if (bucketIdOptional.isPresent()) {
+            return bucketIdOptional.get();
+        }
+
+        String bucketIdStr = cursor.getCursorProperties().get(bucketState.bucketKey());
+        long bucketId = Long.parseLong(bucketIdStr);
+        bucketState.setBucketId(bucketId);
+        return bucketId;
+    }
+
+    private BucketState createBucket(long startLedgerId, long endLedgerId) {
+        BucketState bucketState = new BucketState(startLedgerId, endLedgerId);
+        immutableBuckets.put(Range.closed(startLedgerId, endLedgerId), bucketState);
+        return bucketState;
+    }
+
+    private CompletableFuture<Long> asyncSaveBucketSnapshot(
+            BucketState bucketState, SnapshotMetadata snapshotMetadata,
+            List<SnapshotSegment> bucketSnapshotSegments) {
+
+        return bucketSnapshotStorage.createBucketSnapshot(snapshotMetadata, bucketSnapshotSegments)
+                .thenCompose(newBucketId -> {
+                    bucketState.setBucketId(newBucketId);
+                    String bucketKey = bucketState.bucketKey();
+                    return putBucketKeyId(bucketKey, newBucketId).exceptionally(ex -> {
+                        log.warn("Failed to record bucketId to cursor property, bucketKey: {}", bucketKey);
+                        return null;
```

Comment:
Maybe we can easily return `newBucketId` here.

Reply:
In the `exceptionally`? this only is an exceptional case.

Reply:
Oh, I got it.